File: DSA/VivaswanSingh_590023022/Day_94/q2.c
#include <stdio.h>
#include <stdlib.h>

int compare(const void *a, const void *b)
{
    return (*(int *)a - *(int *)b);
}

int minCost(int n, int cuts[], int cutsSize)
{
    int m = cutsSize + 2;

    int points[m];

    points[0] = 0;
    points[m - 1] = n;

    for (int i = 0; i < cutsSize; i++)
        points[i + 1] = cuts[i];

    qsort(points, m, sizeof(int), compare);

    int dp[m][m];

    for (int i = 0; i < m; i++)
        for (int j = 0; j < m; j++)
            dp[i][j] = 0;

    for (int len = 2; len < m; len++)
    {
        for (int i = 0; i + len < m; i++)
        {
            int j = i + len;
            dp[i][j] = 1000000000;

            for (int k = i + 1; k < j; k++)
            {
                int cost = dp[i][k] + dp[k][j] + points[j] - points[i];

                if (cost < dp[i][j])
                    dp[i][j] = cost;
            }
        }
    }

    return dp[0][m - 1];
}

int main()
{
    int n = 7;

    int cuts[] = {1, 3, 4, 5};
    int cutsSize = sizeof(cuts) / sizeof(cuts[0]);

    int result = minCost(n, cuts, cutsSize);

    printf("Minimum cost: %d\n", result);

    return 0;
}

File: DSA/vaishnavi_joshi_590016947/Day94/Q2.java
import java.util.*;

class Solution {
    public int minCost(int n, int[] cuts) {

        int m = cuts.length;

        // Add 0 and n
        int[] points = new int[m + 2];

        points[0] = 0;
        points[m + 1] = n;

        for (int i = 0; i < m; i++) {
            points[i + 1] = cuts[i];
        }

        // Sort cut positions
        Arrays.sort(points);

        // dp[i][j] = minimum cost to cut between points[i] and points[j]
        int[][] dp = new int[m + 2][m + 2];

        // Length of interval between i and j
        for (int len = 2; len < m + 2; len++) {

            for (int i = 0; i + len < m + 2; i++) {

                int j = i + len;

                dp[i][j] = Integer.MAX_VALUE;

                // Try every possible first cut
                for (int k = i + 1; k < j; k++) {

                    int cost = points[j] - points[i]
                             + dp[i][k]
                             + dp[k][j];

                    dp[i][j] = Math.min(dp[i][j], cost);
                }
            }
        }

        return dp[0][m + 1];
    }
}